File: 0974-subarray-sums-divisible-by-k/0974-subarray-sums-divisible-by-k.py
from collections import defaultdict

class Solution:
    def subarraysDivByK(self, nums: List[int], k: int) -> int:
        result = defaultdict(int)
        answer = 0
        _sum = 0
        result[0] = 1
    
        for num in nums:
            _sum = (_sum + num % k + k) % k
            answer += result[_sum]
            result[_sum] += 1
            
        return answer
            
        

File: 0974-subarray-sums-divisible-by-k/0974-subarray-sums-divisible-by-k.rs
use std::collections::HashMap;

impl Solution {
    pub fn subarrays_div_by_k(nums: Vec<i32>, k: i32) -> i32 {
        let mut cache: HashMap<i32, i32> = HashMap::new();
        
        cache.insert(0, 1);
        
        let mut sum = 0;
        let mut answer = 0;
        
        for num in nums {
            sum = (sum + num % k + k) % k;
            if cache.contains_key(&sum) {
                answer += cache[&sum];
            }
            cache.entry(sum).and_modify(|x| *x += 1).or_insert(1);         
        }
        answer
    }
}
